### PR TITLE
Websockets server compression

### PR DESCRIPTION
**Pull request by disparlure**
Hello,
I've added support for websockets compression for server and client. Server support is enabled by default (but the client must request it). Client side must be specifically enabled in options. Please let me know if there's anything that needs to be changed/added to make this complete. Thanks.

### CODE REVIEW

File: pom.xml [hunk 1]
```diff
@@ -23,7 +23,7 @@
   </parent>
 
   <artifactId>vertx-core</artifactId>
-  <version>3.5.0-SNAPSHOT</version>
+  <version>3.5.0-SNAPSHOT-CMP</version>
```
Comment (vietj) [inline pom.xml]: is it intentional ?

File: pom.xml [hunk 2]
```diff
@@ -689,4 +689,5 @@
     </profile>
   </profiles>
 
+  <description>Updated temporary instance that uses websockets compresson.</description>
```
Comment (vietj) [inline pom.xml]: is it intentional ?

File: src/main/java/io/vertx/core/http/HttpClientOptions.java [hunk 3]
```diff
@@ -154,6 +154,36 @@
    * Default initial buffer size for HttpObjectDecoder = 128 bytes
    */
   public static final int DEFAULT_DECODER_INITIAL_BUFFER_SIZE = 128;
+  
+  /**
+   * Default offer WebSocket Deflate Frame compression
```
Comment (vietj) [inline src/main/java/io/vertx/core/http/HttpClientOptions.java]: please follow the convention of other constant comments

File: src/main/java/io/vertx/core/http/HttpServerOptions.java [hunk 4]
```diff
@@ -681,6 +721,96 @@ public HttpServerOptions setDecoderInitialBufferSize(int decoderInitialBufferSiz
     return this;
   }
   
+  /**
+   * Enable or disable support for WebSocket Defalte Frame compression
+   * @param deflateCompressionSupported
+   * @return a reference to this, so the API can be used fluently
+   */
+  public HttpServerOptions setIsEnabledWebsocketFrameDeflateCompression (boolean deflateCompressionSupported) {
```
Comment (vietj) [inline src/main/java/io/vertx/core/http/HttpServerOptions.java]: shouldn't it rather be `setWebsocketFrameDeflateCompressionSupported`, same for other getter/setter

File: src/test/java/io/vertx/test/core/WebsocketTest.java [hunk 5]
```diff
@@ -931,6 +932,120 @@ public void run() {
...
+		          }
+			  });
+		  });
+	  });
+	  await();
+  }
+  
+  @Test
+  // Test normal negotiation of websocket compression
+  public void testNormalWSPermessageDefalteCompressionNegotiation() throws Exception {
```
Comment (vietj) [inline src/test/java/io/vertx/test/core/WebsocketTest.java]: typo Defalte -> Deflate

File: src/main/java/io/vertx/core/http/HttpClientOptions.java [hunk 6]
```diff
@@ -943,6 +993,124 @@ public HttpClientOptions setLogActivity(boolean logEnabled) {
   }
 
   /**
+   * Set option to offer Deflate Frame websocket compression
+   * @param tryDeflateFrame
+   * @return  a reference to this, so the API can be used fluently
+   */
+ public HttpClientOptions setTryWebsocketDefalteFrameCompression(boolean tryDeflateFrame)
```
Comment (vietj) [inline src/main/java/io/vertx/core/http/HttpClientOptions.java]: typo Defalte -> Deflate

File: src/test/java/io/vertx/test/core/WebsocketTest.java [hunk 7]
```diff
@@ -931,6 +932,120 @@ public void run() {
     await();
   }
 
+  @Test
+  // Test normal negotiation of websocket compression
+  public void testNormalWSDeflateFrameCompressionNegotiation() throws Exception {
+	  String path = "/some/path";
```
Comment (vietj) [inline src/test/java/io/vertx/test/core/WebsocketTest.java]: it seems that indent is not 2 spaces

File: pom.xml (hunk 1 above)
Comment (disparlure) [inline pom.xml]: No, I was using this marker for my local branch to differentiate with the actual current version. Will back it out.

Comment (disparlure) [inline pom.xml]: No, I had been using that marker to differentiate the modified version from the master in my fork. I'll remove it.

**Comment by disparlure** [comment]
Inadvertently committed with wrong account set. Have gone back and re-signed ECA with the account that I used for the commits.

File: src/main/java/io/vertx/core/http/HttpClientOptions.java [hunk 8]
```diff
@@ -943,6 +993,124 @@ public HttpClientOptions setLogActivity(boolean logEnabled) {
   }
 
   /**
+   * Set option to offer Deflate Frame websocket compression
+   * @param tryDeflateFrame
+   * @return  a reference to this, so the API can be used fluently
+   */
+ public HttpClientOptions setTryWebsocketDeflateFrameCompression(boolean tryDeflateFrame)
+ {
+	 this.websocketTryUseDeflateFrame = tryDeflateFrame;
```
Comment (vietj) [inline src/main/java/io/vertx/core/http/HttpClientOptions.java]: the field should be named `tryWebsocketDeflateFrameCompression`

File: src/main/java/io/vertx/core/http/HttpClientOptions.java [hunk 9]
```diff
@@ -943,6 +993,124 @@ public HttpClientOptions setLogActivity(boolean logEnabled) {
...
+	 return this;
+ }
+ 
+ /** 
+  * 
+  * @return Whether Deflate Frame websocket compression will be offered
+  */
+ public boolean isTryWebsocketDefalteFrameCompressionEnabled()
+ {
+	 return this.websocketTryUseDeflateFrame;
```
Comment (vietj) [inline src/main/java/io/vertx/core/http/HttpClientOptions.java]: the field should be named `tryWebsocketDeflateFrameCompression `

File: src/main/java/io/vertx/core/http/HttpClientOptions.java [hunk 10]
```diff
@@ -943,6 +993,124 @@ public HttpClientOptions setLogActivity(boolean logEnabled) {
...
+ {
+	 this.websocketTryUseDeflateFrame = tryDeflateFrame;
+	 return this;
+ }
+ 
+ /** 
+  * 
+  * @return Whether Deflate Frame websocket compression will be offered
+  */
+ public boolean isTryWebsocketDefalteFrameCompressionEnabled()
```
Comment (vietj) [inline src/main/java/io/vertx/core/http/HttpClientOptions.java]: this should be named `isTryWebsocketDeflateFrameCompression`

File: src/main/java/io/vertx/core/http/HttpClientOptions.java [hunk 11]
```diff
@@ -943,6 +993,124 @@ public HttpClientOptions setLogActivity(boolean logEnabled) {
...
+ }
+ 
+/**
+ * Set option to offer Permessage Deflate websocket compression
+ * @param tryPermessageDeflate
+ * @return  a reference to this, so the API can be used fluently
+ */
+ public HttpClientOptions setTryWebsocketTryPermessageDefalteCompression(boolean tryPermessageDeflate)
+ {
+	 this.websocketTryUsePermessageDeflate = tryPermessageDeflate;
```
Comment (vietj) [inline src/main/java/io/vertx/core/http/HttpClientOptions.java]: this should be named `tryWebsocketPerMessageDeflatCompression`

File: src/main/java/io/vertx/core/http/HttpClientOptions.java [hunk 12]
```diff
@@ -943,6 +993,124 @@ public HttpClientOptions setLogActivity(boolean logEnabled) {
...
+ {
+	 return this.websocketTryUseDeflateFrame;
+ }
+ 
+/**
+ * Set option to offer Permessage Deflate websocket compression
+ * @param tryPermessageDeflate
+ * @return  a reference to this, so the API can be used fluently
+ */
+ public HttpClientOptions setTryWebsocketTryPermessageDefalteCompression(boolean tryPermessageDeflate)
```
Comment (vietj) [inline src/main/java/io/vertx/core/http/HttpClientOptions.java]: this should be named `setTryWebsocketDeflateFrameCompression`

Comment (disparlure) [inline src/main/java/io/vertx/core/http/HttpClientOptions.java]: Are you sure? This is the function to set permessage deflate. Also, in the previous comment for the deflate frame compression setter, you recommended removing the 'set' prefix from the name.

**Comment by CodingFabian** [comment]
anything we could contribute to get this rolling?

**Comment by disparlure** [comment]
What needs to be done now? I've been waiting on further review/information.

**Comment by vietj** [comment]
I will do another review soon

**Comment by disparlure** [comment]
Any update on this?

**Comment by vietj** [comment]
I will have a look tomorrow, I've been much busy with Netty upgrade and other networking improvements

**Comment by vietj** [comment]
I've started to look at it, I will provide more feedback soon.

**Comment by vietj** [comment]
here is another review pass:

- is DeflateUseWebKitName really necessary ? it seems to be only used on mobile Safari client, so it's not really useful in Vert.x HttpClient. So the only purpose would be for HttpClient to emulate Safari behavior. Furthermore it seems to be depracated : https://bugs.webkit.org/show_bug.cgi?id=115504 . On the server it is handled automatically without configuration.
- the NoContext options don't seem much useful as they limit the gain of compression, I would recommend for a first iteration of this feature to keep it aside and leave it enabled by default. If there is demand we can still add it later.
- on the client, some options should be renamed to be more clear so we only keep
   - tryUsePerMessageWebsocketCompression (renamed)
   - tryUsePerFrameWebsocketCompression (renamed)
   - 
- on the server, some options should be renamed to be more clear
   - perMessageWebsocketCompressionSupported (renamed)
   - perFrameWebsocketCompressionSupported (renamed)
   - websocketCompressionLevel

In addition you need to rebase the PR on the latest version of Vert.x

I'll make another pass after this.

**Comment by vietj** [comment]
for those interested, here is a primer on websocket compression : https://www.igvita.com/2013/11/27/configuring-and-optimizing-websocket-compression/#selective

it seems interesting to support specially in the server case because now browsers seems to implement it, specially the per message case.

in case of event bus bridge, it might be interesting to support as well.

**Comment by disparlure** [comment]
I pulled out the DeflateUseWebKitName option as obsolete, but have left the NoContext options in; they are accessed by the application we're building on top of Vert.x.

**Comment by vietj** [comment]
you need to sign the Eclipse Contributor Agreement @disparlure

**Comment by disparlure** [comment]
@vietj  I've signed the ECA using both of the credentials that I've used for this project - it doesn't seem to take.

**Comment by vietj** [comment]
sign your commits as well, I think now you should make a new clean pr

**Comment by disparlure** [comment]
Won't there still be issues with the unsigned commits? I've signed the more recent ones, but the first few are unsigned.

**Comment by dgoetsch** [comment]
You could rebasend and squash to one commit :)

**Comment by vietj** [comment]
all commits should be signed wether you squash them or not

**Comment by vietj** [comment]
see in the PR there is a "ip-validation" check that prevents merging

**Comment by disparlure** [comment]
Closing to create new clean PR

**Comment by pvub** [comment]
@disparlure Thanks for making this change. But I do not see this available on the latest 3.5.0 build. Any idea when will this be available? I am not sure if there is another Change Request for this. 
@vietj I really need a way to enable 'permessage-deflate' on both the Client and Server sides, and this change would be immensely helpful. 

Thanks in advance.

**Comment by vietj** [comment]
Sure thing in 3.5.1

**Comment by pvub** [comment]
@vietj Thank you. Any idea when will 3.5.1 be available?

**Comment by CodingFabian** [comment]
also leaving this here - we cleaned up the great contribution of @disparlure in #2165 pending to be merged.

**Pull request closed without merging**